File: examples/agentkit/crewai/agent.py
"""
CrewAI agent with Scalekit-authenticated Gmail tools via MCP.
Mirrors: /agentkit/examples/crewai/
Run: python agent.py (from examples/agentkit/, using crewai-venv)
"""
import os
from typing import Any, Optional
import scalekit.client
from dotenv import load_dotenv

load_dotenv()

# Patch CrewAI's schema converter to handle nullable types (['string', 'null'] etc.)
# before importing MCPServerAdapter, which triggers tool schema parsing on import.
import crewai.utilities.pydantic_schema_utils as _schema_mod
_orig = _schema_mod._json_schema_to_pydantic_type

def _patched(json_schema: dict[str, Any], root_schema: dict[str, Any], **kwargs: Any) -> Any:
    type_ = json_schema.get("type")
    if isinstance(type_, list):
        non_null = [t for t in type_ if t != "null"]
        has_null = "null" in type_
        inner = _orig({**json_schema, "type": non_null[0] if non_null else "string"}, root_schema, **kwargs)
        return Optional[inner] if has_null else inner  # type: ignore[return-value]
    return _orig(json_schema, root_schema, **kwargs)

_schema_mod._json_schema_to_pydantic_type = _patched

from crewai import Agent, Task, Crew, LLM
from crewai_tools import MCPServerAdapter

scalekit_client = scalekit.client.ScalekitClient(
    client_id=os.getenv("SCALEKIT_CLIENT_ID"),
    client_secret=os.getenv("SCALEKIT_CLIENT_SECRET"),
    env_url=os.getenv("SCALEKIT_ENV_URL"),
)
actions = scalekit_client.actions

# Connect user to Gmail
response = actions.get_or_create_connected_account(
    connection_name="gmail",
    identifier="docs-example",
)
if response.connected_account.status != "ACTIVE":
    link = actions.get_authorization_link(connection_name="gmail", identifier="docs-example")
    print("Authorize Gmail:", link.link)
    input("Press Enter after authorizing...")

# Get a per-user MCP URL
inst_response = actions.mcp.ensure_instance(
    config_name="gmail-docs-example",
    user_identifier="docs-example",
)
mcp_url = inst_response.instance.url

with MCPServerAdapter({"url": mcp_url, "transport": "streamable-http"}) as mcp_tools:
    agent = Agent(
        role="Email Assistant",
        goal="Fetch and summarize the user's unread emails",
        backstory="You are a helpful assistant with access to the user's Gmail inbox.",
        tools=mcp_tools,
        llm=LLM(model="claude-sonnet-4-6", base_url=os.getenv("OPENAI_BASE_URL"), api_key=os.getenv("OPENAI_API_KEY")),
        verbose=True,
    )

    task = Task(
        description="Fetch the last 5 unread emails and provide a brief summary of each.",
        expected_output="A list of 5 unread emails with subject, sender, and a one-sentence summary.",
        agent=agent,
    )

    result = Crew(agents=[agent], tasks=[task]).kickoff()
    print(result)

File: examples/agentkit/mastra/agent.ts
/**
 * Mastra agent with Scalekit-authenticated Gmail tools via MCP.
 * Mirrors: /agentkit/examples/mastra/
 * Run: tsx mastra/agent.ts (from examples/agentkit/)
 *
 * Prerequisite: set SCALEKIT_MCP_URL in .env
 * Generate it via Python: actions.mcp.ensure_instance(config_name="...", user_identifier="docs-example")
 */
import { Agent } from '@mastra/core/agent'
import { MCPClient } from '@mastra/mcp'
import { createOpenAI } from '@ai-sdk/openai'
import 'dotenv/config'

const mcpUrl = process.env.SCALEKIT_MCP_URL
if (!mcpUrl) {
  console.error(
    'SCALEKIT_MCP_URL not set. Generate it with actions.mcp.ensure_instance() and add to .env',
  )
  process.exit(1)
}

const mcp = new MCPClient({
  servers: {
    scalekit: { url: new URL(mcpUrl) },
  },
})

const toolsets = await mcp.listToolsets()
const tools = Object.values(toolsets).reduce((acc, ts) => ({ ...acc, ...ts }), {})

const openaiProvider = createOpenAI({
  baseURL: process.env.OPENAI_BASE_URL,
  apiKey: process.env.OPENAI_API_KEY,
})

const agent = new Agent({
  name: 'gmail_assistant',
  instructions: 'You are a helpful Gmail assistant.',
  model: openaiProvider.chat('claude-sonnet-4-6'),
  tools,
})

const result = await agent.generate('Fetch my last 5 unread emails and summarize them')
console.log(result.text)

await mcp.disconnect()
